File: _posts/design-patterns/visitor/src/Visitable.java
public class Visitable {
  public static void main(String[] args) {
    Expr e = new UnaryExpr("-", new BinaryExpr("+", new Literal(2), new Literal(3)));
    System.out.println(e.accept(new EvalVisitor()));
  }
}

interface Expr {
  default double accept(Visitor v) {
    v.visit(this);
  }
}

class BinaryExpr implements Expr {
  Expr lhs;
  Expr rhs;
  String op;

  BinaryExpr(String op, Expr lhs, Expr rhs) {
    this.op = op;
    this.lhs = lhs;
    this.rhs = rhs;
  }
}

class UnaryExpr implements Expr {
  String op;
  Expr x;

  UnaryExpr(String op, Expr x) {
    this.op = op;
    this.x = x;
  }
}

class Literal implements Expr {
  double v;

  Literal(double v) {
    this.v = v;
  }
}

interface Visitor {
  double visit(BinaryExpr e);
  double visit(UnaryExpr e);
  double visit(Literal e);
}

class EvalVisitor implements Visitor {
  public double visit(BinaryExpr e) {
    switch (e.op) {
      case "+":
        return e.lhs.accept(this) + e.rhs.accept(this);
    }
  }
  public double visit(UnaryExpr e) {
    switch (e.op) {
      case "-":
        return -e.x.accept(this);
    }
  }
  public double visit(Literal e) {
    return e.v;
  }
}

File: assets/design-patterns/visitor/src/visitable.rb
#!/usr/bin/env ruby

class Expr
  def accept v
    method = "visit_" + self.class.name.split(/([A-Z][a-z]+)/).filter { |s| not s.empty? }.map { |s| s.downcase }.inject { |accum, p|
 accum + "_" + p }
    v.send method.to_sym, self
  end
end

class BinaryExpr < Expr
  attr_reader :op, :lhs, :rhs

  def initialize op, lhs, rhs
    @op = op
    @lhs = lhs
    @rhs = rhs
  end
end

class UnaryExpr < Expr
  attr_reader :op, :x

  def initialize op, x
    @op = op
    @x = x
  end
end

class Literal < Expr
  attr_reader :v

  def initialize v
    @v = v
  end
end

class EvalVisitor
  def visit_binary_expr e
    case e.op
    when "+"
      return e.lhs.accept(self) + e.rhs.accept(self)
    end
  end

  def visit_unary_expr e
    case e.op
    when "-"
      return -e.x.accept(self)
    end
  end

  def visit_literal e
    e.v
  end
end

p UnaryExpr.new(
  "-",
  BinaryExpr.new(
    "+",
    Literal.new(2),
    Literal.new(3),
  )
).accept(EvalVisitor.new)
